File: DemoApp/DemoApp/InterstitialsDetailViewController.m
//
//  InterstitialsDetailViewController.m
//  DemoApp
//
//  Created by Glispa GmbH on 04/12/15.
//  Copyright © 2015 Glispa GmbH All rights reserved.
//

#import "InterstitialsDetailViewController.h"
#import "AppControllerManagement.h"

@interface InterstitialsDetailViewController ()

@property(strong, nonatomic) AVOInterstitial *interstitial;
@end

@implementation InterstitialsDetailViewController

- (void)viewDidLoad {
    [super viewDidLoad];
}


#pragma mark - Actions

- (IBAction)loadClicked:(id)sender {
    __weak typeof(self) weakSelf = self;
    [AvocarrotSDK.sharedSDK loadInterstitialWithAdUnitId:self.adUnitId
                                                 success:^(AVOInterstitial *interstitial) {
                                                     __strong __typeof__(self) sSelf = weakSelf;
                                                     sSelf.interstitial = interstitial;
                                                     NSLog(@"Interstitial has loaded!");
                                                 } failure:^(AVOError * _Nonnull error) {
                                                     NSLog(@"Interstitial loading error: %@", [error localizedDescription]);
                                                 }];
}

- (IBAction)showClicked:(id)sender {
    [self.interstitial showFromViewController:self];
}

- (IBAction)loadAndShowClicked:(id)sender {
    __weak typeof(self) weakSelf = self;
    [AvocarrotSDK.sharedSDK loadInterstitialWithAdUnitId:self.adUnitId
                                              success:^(AVOInterstitial *interstitial) {
                                                  __strong __typeof__(self) sSelf = weakSelf;
                                                  sSelf.interstitial = interstitial;
                                                  [interstitial showFromViewController:sSelf];
                                                  NSLog(@"Interstitial has loaded! And it will be shown...");
                                              } failure:^(AVOError * _Nonnull error) {
                                                  NSLog(@"Interstitial loading error: %@", [error localizedDescription]);
                                              }];
}

- (IBAction)loadAndShowWithDelayClicked:(id)sender {
    __weak typeof(self) weakSelf = self;
    [AvocarrotSDK.sharedSDK loadInterstitialAndShowWithDelayWithAdUnitId:self.adUnitId
                                                       forViewController:self
                                                                 success:^(AVOInterstitial *interstitial) {
                                                                     __strong __typeof__(self) sSelf = weakSelf;
                                                                     sSelf.interstitial = interstitial;
                                                                     NSLog(@"Interstitial has loaded! And it will be shown soon...");
                                                                 }
                                                                 failure:^(AVOError * _Nonnull error) {
                                                                     NSLog(@"Interstitial loading error: %@", [error localizedDescription]);
                                                                 }];
}

@end


@interface InterstitialsDetailViewController (AppManagement) <AppControllerManagement>

@end

@implementation InterstitialsDetailViewController (AppManagement)

- (void)applyUserData:(NSDictionary <NSString *, NSObject *> *)userData {
    if ([userData[@"adUnitId"] isKindOfClass:[NSString class]]) {
        self.adUnitId = (NSString *)userData[@"adUnitId"];
    }
}

@end

File: DemoApp/DemoAppSwift/InterstitialsDetailViewController.swift
//
//  InterstitialsDetailViewController.swift
//  DemoApp
//
//  Created by Glispa GmbH on 19/05/16.
//  Copyright © 2016 Glispa GmbH. All rights reserved.
//

import UIKit
import AvocarrotInterstitial

class InterstitialsDetailViewController: UIViewController {

    @IBOutlet weak var loadButton: UIButton!
    @IBOutlet weak var showButton: UIButton!
    @IBOutlet weak var loadAndShowButton: UIButton!
    @IBOutlet weak var loadAndShowWithDelayButton: UIButton!

    public var adUnitId: String!

    private var interstitial: AVOInterstitial?

    override func viewDidLoad() {
        super.viewDidLoad()
        showButton.isEnabled = false
    }

    // MARK: - Actions

    @IBAction func loadClicked(_ sender: UIButton) {
        showButton.isEnabled = false
        AvocarrotSDK.shared.loadInterstitial(withAdUnitId: adUnitId, success: { [weak self] (interstitial) in
                                            guard let sSelf = self else {return}
                                            sSelf.interstitial = interstitial
                                            sSelf.showButton.isEnabled = true
            
                                            interstitial.onClick{
                                                print("The intestitial is clicked")
                                            }
            
                                            print("Interstitial has loaded!")
                                           }, failure: { (error) in
                                            print("Interstitial loading error: \(error.localizedDescription)")
                                           })
    }

    @IBAction func showClicked(_ sender: UIButton) {
        interstitial?.show(from: self)
        showButton.isEnabled = false
    }

    @IBAction func loadAndShowClicked(_ sender: UIButton) {
        showButton.isEnabled = false
        AvocarrotSDK.shared.loadInterstitial(withAdUnitId: adUnitId,
                                              success: { [weak self] (interstitial) in
                                                guard let sSelf = self else {return}
                                                sSelf.interstitial = interstitial
                                                
                                                interstitial.onClick{
                                                    print("The intestitial is clicked")
                                                }
                                                
                                                interstitial.show(from: sSelf)
                                                print("Interstitial has loaded! And it will be shown...")
                                              }, failure: { (error) in
                                                print("Interstitial loading error: \(error.localizedDescription)")
                                              })

    }

}

extension InterstitialsDetailViewController: AppControllerManagement {
    func applyUserData(_ userData: [String:Any]) {
        if let adId = userData["adUnitId"] as? String {
            adUnitId = adId
        }
    }
}
